### PR TITLE
[1.8 release] Switch to the new datasets in torchtext 0.9.0 release - text classification tutorial

### PR DESCRIPTION
**Pull request by zhangguanheng66**
In torchtext 0.9.0 release, we will include the raw text datasets as beta release. Update the text classification tutorial with the new torchtext library.

This PR should be tested against pytorch 1.8.0 rc and torchtext 0.9.0 rc.

### CODE REVIEW

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 1]
```diff
@@ -1,66 +1,131 @@
...
+
+
+
+######################################################################
+# Generate data batch and iterator 
+# --------------------------------
+#
+# `torch.utils.data.DataLoader <https://pytorch.org/docs/stable/data.html?highlight=dataloader#torch.utils.data.DataLoader>`__
+# is recommended for PyTorch users (a tutorial is `here <https://pytorch.org/tutorials/beginner/data_loading_tutorial.html>`__).
+# It works with a map-style dataset that implements the ``getitem()`` and ``len()`` protocols, and represents a map from indices/keys to data samples. It also works with an iterable datasets with the shuffle argumnet of ``False``.
```
Comment (cpuhrsch) [inline beginner_source/text_sentiment_ngrams_tutorial.py:98]: nit: argumnet -> argument

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 2]
```diff
@@ -1,66 +1,131 @@
...
+#
+# Before sending to the model, ``collate_fn`` function works on a batch of samples generated from ``DataLoader``. The input to ``collat_fn`` is a batch of data with the batch size in ``DataLoader``, and ``collate_fn`` processes them according to the data processing pipelines declared previouly. Pay attention here and make sure that ``collate_fn`` is declared as a top level def. This ensures that the function is available in each worker.
```
Comment (cpuhrsch) [inline beginner_source/text_sentiment_ngrams_tutorial.py:100]: nit: collat_fn -> collate_fn

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 3]
```diff
@@ -1,66 +1,131 @@
...
+         processed_text = torch.tensor(text_pipeline(_text), dtype=torch.int64)
+         text_list.append(processed_text)
+         offsets.append(processed_text.size(0))
+    label_list = torch.tensor(label_list, dtype=torch.int64)
+    offsets = torch.tensor(offsets[:-1]).cumsum(dim=0)
+    text_list = torch.cat(text_list)
+    return label_list.to(device), text_list.to(device), offsets.to(device)    
+
+train_iter = AG_NEWS(split='train')
+dataloader = DataLoader(list(train_iter), batch_size=8, shuffle=True, collate_fn=collate_batch)
```
Comment (cpuhrsch) [inline beginner_source/text_sentiment_ngrams_tutorial.py:121]: I think we should split out ```list(train_iter)``` into its separate line and address it. Here the fact that the dataset is small makes it feasible to just materialize it into a list, but in general someone might want to use DataLoader for multiple streams (e.g. to apply tokenization and Vocab concurrently). Here the ```offset``` argument can help.

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 4]
```diff
@@ -103,51 +168,14 @@ def forward(self, text, offsets):
...
-    offsets = [0] + [len(entry) for entry in text]
-    # torch.Tensor.cumsum returns the cumulative sum
-    # of elements in the dimension dim.
-    # torch.Tensor([1.0, 2.0, 3.0]).cumsum(dim=0)
-
-    offsets = torch.tensor(offsets[:-1]).cumsum(dim=0)
-    text = torch.cat(text)
-    return text, offsets, label
+train_iter = AG_NEWS(split='train')
+num_class = len(set([label for (label, text) in train_iter]))
```
Comment (cpuhrsch) [inline beginner_source/text_sentiment_ngrams_tutorial.py:175]: Here we're materializing the dataset again, but this already happened earlier in the context of DataLoader. We can just assign ```list(train_iter)``` to a variable to avoid this. We should probably also add the number of labels to our dataset documentation, which would be much more efficient to use than this. I'll add this as a task.

File: beginner_source/text_sentiment_ngrams_tutorial.py (hunk 1 above)
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py]: Fixed

File: beginner_source/text_sentiment_ngrams_tutorial.py (hunk 3 above)
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py]: Fixed the tutorial by passing an IterableDataset as an example.

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 5]
```diff
@@ -1,66 +1,135 @@
 """
-Text Classification with TorchText
+Text classification with the new torchtext library
```
Comment (cpuhrsch) [inline beginner_source/text_sentiment_ngrams_tutorial.py:2]: I'd remove the word "new", because this is now the current torchtext library. So seeing "new" in a stable tutorial might be confusing to a user.

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 6]
```diff
@@ -2,7 +2,7 @@
 Text classification with the torchtext library
 ==================================
 
-In this tutorial, we will show how to use the new torchtext library to build the dataset for the text classification analysis. In the nightly release of the torchtext library, we provide a few prototype building blocks for data processing. Users will have the flexibility to
+In this tutorial, we will show how to use the new torchtext library to build the dataset for the text classification analysis. Users will have the flexibility to
```
Comment (cpuhrsch) [inline beginner_source/text_sentiment_ngrams_tutorial.py:5]: We don't need to say "new" torchtext library anymore, because the datasets are now part of the top folder.

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 7]
```diff
@@ -156,188 +186,211 @@ def generate_batch(batch):
...
-    scheduler.step()
+        total_acc += (predited_label.argmax(1) == label).sum().item()
+        total_count += label.size(0)
+        if idx % log_interval == 0 and idx > 0:
+            elapsed = time.time() - start_time
+            print('| epoch {:3d} | {:5d}/{:5d} batches '
+                  '| accuracy {:8.3f}'.format(epoch, idx, len(dataloader),
+                                              total_acc/total_count))
+            total_acc, total_count = 0, 0
+            start_time = time.time()
```
Comment (parmeet) [inline beginner_source/text_sentiment_ngrams_tutorial.py:212]: unused variable?

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 8]
```diff
@@ -156,188 +186,211 @@ def generate_batch(batch):
...
-# `torch.utils.data.DataLoader <https://pytorch.org/docs/stable/data.html?highlight=dataloader#torch.utils.data.DataLoader>`__
-# is recommended for PyTorch users, and it makes data loading in parallel
-# easily (a tutorial is
-# `here <https://pytorch.org/tutorials/beginner/data_loading_tutorial.html>`__).
-# We use ``DataLoader`` here to load AG_NEWS datasets and send it to the
-# model for training/validation.
-#
-
-from torch.utils.data import DataLoader
+import time
```
Comment (parmeet) [inline beginner_source/text_sentiment_ngrams_tutorial.py:189]: perhaps can be imported in next code snippet as it might not be used here?

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 9]
```diff
@@ -69,11 +136,12 @@
...
-import torch.nn.functional as F
-class TextSentiment(nn.Module):
+from torch import nn
+
+class TextClassificationModel(nn.Module):
+
     def __init__(self, vocab_size, embed_dim, num_class):
-        super().__init__()
+        super(TextClassificationModel, self).__init__()
         self.embedding = nn.EmbeddingBag(vocab_size, embed_dim, sparse=True)
```
[the remark this answers is not in the document]
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py:145]: I revise the context with "the default mode of mean".

File: beginner_source/text_sentiment_ngrams_tutorial.py (hunk 8 above)
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py:189]: It's used in L195.

File: beginner_source/text_sentiment_ngrams_tutorial.py (hunk 7 above)
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py:212]: Here we reset the start_time variable to have the new time period.

File: beginner_source/text_sentiment_ngrams_tutorial.py (hunk 6 above)
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py]: Fixed.

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 10]
```diff
@@ -1,66 +1,133 @@
...
 # Define the model
 # ----------------
 #
-# The model is composed of the
-# `EmbeddingBag <https://pytorch.org/docs/stable/nn.html?highlight=embeddingbag#torch.nn.EmbeddingBag>`__
-# layer and the linear layer (see the figure below). ``nn.EmbeddingBag``
-# computes the mean value of a “bag” of embeddings. The text entries here
-# have different lengths. ``nn.EmbeddingBag`` requires no padding here
-# since the text lengths are saved in offsets.
+# The model is composed of the `nn.EmbeddingBag <https://pytorch.org/docs/stable/nn.html?highlight=embeddingbag#torch.nn.EmbeddingBag>`__ layer plus a linear layer for the classification purpose. ``nn.EmbeddingBag`` computes the mean value of a “bag” of embeddings. Although the text entries here have different lengths, nn.EmbeddingBag module requires no padding here since the text lengths are saved in offsets.
```
[the remark this answers is not in the document]
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py]: Update the context and explicitly say the default mode of mean.

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 11]
```diff
@@ -156,188 +186,211 @@ def generate_batch(batch):
...
-# `here <https://pytorch.org/tutorials/beginner/data_loading_tutorial.html>`__).
-# We use ``DataLoader`` here to load AG_NEWS datasets and send it to the
-# model for training/validation.
-#
-
-from torch.utils.data import DataLoader
+import time
 
-def train_func(sub_train_):
+def train(model, dataloader):
```
[the remark this answers is not in the document]
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py]: OK, to keep them consistent, I use global variables in the functions.

File: beginner_source/text_sentiment_ngrams_tutorial.py [hunk 12]
```diff
@@ -156,188 +186,211 @@ def generate_batch(batch):
...
+            print('| epoch {:3d} | {:5d}/{:5d} batches '
+                  '| accuracy {:8.3f}'.format(epoch, idx, len(dataloader),
+                                              total_acc/total_count))
+            total_acc, total_count = 0, 0
+            start_time = time.time()
+
+def evaluate(model, dataloader):
+    model.eval()
+    total_acc, total_count = 0, 0
+    ans_pred_tokens_samples = []
```
[the remark this answers is not in the document]
Comment (zhangguanheng66) [reply beginner_source/text_sentiment_ngrams_tutorial.py]: Removed

**Review by parmeet — APPROVED** [review]
